core/src/main/scala/kafka/log/Log.scala

```diff
@@ -392,18 +392,27 @@ class Log(@volatile var dir: File,
         // if it's a log file, load the corresponding log segment
         val baseOffset = offsetFromFile(file)
         val timeIndexFileNewlyCreated = !Log.timeIndexFile(dir, baseOffset).exists()
+        val offsetIndexExists = Log.offsetIndexFile(dir, baseOffset).exists()
         val segment = LogSegment.open(dir = dir,
           baseOffset = baseOffset,
           config,
           time = time,
           fileAlreadyExists = true)
-
-        try segment.sanityCheck(timeIndexFileNewlyCreated)
-        catch {
-          case _: NoSuchFileException =>
+        
+        try {
+          // Resize the time index file to 0 if it is newly created.
+          if (timeIndexFileNewlyCreated)
+            segment.timeIndex.resize(0)
+          // Rebuild the index if the index file does not exist
+          if (!offsetIndexExists) {
             error(s"Could not find offset index file corresponding to log file ${segment.log.file.getAbsolutePath}, " +
               "recovering segment and rebuilding index files...")
             recoverSegment(segment)
+          } else if (baseOffset >= recoveryPoint) 
+            // Only sanity check segments above the recovery point
+            segment.sanityCheck()
+        }
+        catch {
```

Comment:
nit: you could start this `catch` block on the previous line

Reply:
Done.